src/AssemblerHttpServer-Alignments.cpp

```diff
@@ -1182,6 +1183,255 @@ void Assembler::computeAllAlignments(
 }
 
 
+void Assembler::sampleReads(vector<OrientedReadId>& sample, uint64_t n){
+    sample.clear();
+
+    while (sample.size() < n) {
+        // Randomly select a read in the read set
+        uint16_t readId = uint16_t(rand() % reads.size());
+
+        // Randomly select an orientation
+        uint16_t strand = uint16_t(rand() % 2);
+
+        const OrientedReadId r(readId, strand);
+
+        sample.push_back(r);
+    }
+}
+
+
+void Assembler::sampleReads(vector<OrientedReadId>& sample, uint64_t n, uint64_t minLength, uint64_t maxLength){
+    sample.clear();
+
+    while (sample.size() < n) {
+        // Randomly select a read in the read set
+        uint16_t readId = uint16_t(rand() % reads.size());
+
+        // Randomly select an orientation
+        uint16_t strand = uint16_t(rand() % 2);
+
+        const OrientedReadId r(readId, strand);
+
+        // Number of raw bases.
+        const auto repeatCounts = readRepeatCounts[readId];
+        uint64_t length = 0;
+        for(const auto repeatCount: repeatCounts) {
+            length += repeatCount;
+        }
+
+        // Only update the sample of reads if this read passes the length criteria
+        if(length >= minLength and length <= maxLength) {
+            sample.push_back(r);
+        }
+    }
+}
+
+
+// Compute alignments on an oriented read against
```

Comment:
For a full human genome, aligning one read with all others takes 40+ seconds on a 128 thread machine. Won't this take minutes for a small sample size of 100?  

If we can't wait that long in a browser, may be we could have this function spawn a thread that generates this data and writes it out to the ShastaRun directory. Even better, I would do this during the main Shasta run and dump the histograms in files for later analysis. Adding 5 minutes to the assembly time seems totally worth it, if the eventual goal is to pick these thresholds automatically anyways. 

What do you think? If you agree, I can help you move this code to the main assembly process.

Reply:
On a large run and depending on alignment parameters, read length distribution, and machine size, this can take several minutes per read. So if we wanted to do it for 50 reads it could take hours. This would be too much time to add to the assembly by default, but we could still do it in parallel for a small number of reads if we can squeeze it in one of our entirely sequential phases. Even in this case, I would still do this optionally and not by default, at least for now.

Reply:
Another alternative would be to trust the LowHash step, and do this thresholding on the reads that have been paired already, instead of 1-to-all.

Reply:
A couple of comments on the histograms. Each bin is labeled with one value (e. g. 0.55) and it is not clear what the bin interval is.Does 0.55 mean "0.55 to 0.60", "0.50 to 0.55", or "0.525 to 0.575"? We have to make this clear, either by adding a second column to characterize each bin (and labeling the two colummns "Min" and "Max" or something similar), or with a comment (I prefer the former solution).

The title of the second histogram is confusing. What does "Aligned Quantity distribution" mean? I imagine it is the number of alignments found for each read. This needs to be clarified.

Don't we also need a histogram for the number of aligned markers?

Reply:
The histogram for number of aligned markers is included, but not shown in the picture because it is long (0-3000). 

Aligned Quantity distribution is the number of alignments found for each read. I agree the name is not clear.

Ok, the binning is easy to do

Reply:
> Another alternative would be to trust the LowHash step, and do this thresholding on the reads that have been paired already, instead of 1-to-all.

I don't see much point in this, because one of the main purposes of doing this is to check that the LowHash step is doing its job.

Reply:
> The histogram for number of aligned markers is included, but not shown in the picture because it is long (0-3000).

You can bin it, like the other hsistograms.

Reply:
> I don't see much point in this, because one of the main purposes of doing this is to check that the LowHash step is doing its job.

It is binned, but it is still long because the interval size needs to be small enough to see the distribution clearly on the lower end.

> I don't see much point in this, because one of the main purposes of doing this is to check that the LowHash step is doing its job.

Hmm I was thinking that the purpose was to remove the need to specify Align.minMarkerCount and Align.minAlignedFraction as parameters in the config. I think it would be fair to estimate the distribution of the alignments given the candidate pairs for this purpose.

Reply:
example:
![image](https://user-images.githubusercontent.com/28764332/88305804-c63d9a80-ccbe-11ea-9581-d697d50c2ff2.png)

Reply:
> Hmm I was thinking that the purpose was to remove the need to specify Align.minMarkerCount and Align.minAlignedFraction as parameters in the config. I think it would be fair to estimate the distribution of the alignments given the candidate pairs for this purpose.

This is a good point. There is a scenario where relying on the stored alignments would be useful.

Reply:
> It is binned, but it is still long because the interval size needs to be small enough to see the distribution clearly on the lower end.

So perhaps use a non-uniform binning. Perhaps uniform in log space?